hazelcast-sql/src/main/java/com/hazelcast/sql/impl/calcite/validate/HazelcastSqlValidator.java

```diff
@@ -17,53 +17,284 @@
+            // Assign DOUBLE type to any standalone floating point literal: the
+            // exact type is inferred later from the context in which the literal
+            // appears.
+
+            derived = HazelcastTypeFactory.INSTANCE.createSqlType(DOUBLE);
+            setKnownAndValidatedNodeType(expression, derived);
+        }
+
+        return derived;
+    }
+
+    @SuppressWarnings({"checkstyle:CyclomaticComplexity", "checkstyle:NPathComplexity"})
+    private RelDataType deriveCastType(RelDataType derived, SqlValidatorScope scope, SqlNode expression) {
+        RelDataType known = knownNodeTypes.get(expression);
+        if (derived == known) {
+            return derived;
+        }
+
+        SqlCall call = (SqlCall) expression;
+        SqlNode operand = call.operand(0);
+        RelDataType from = deriveType(scope, operand);
+
+        RelDataType to = deriveType(scope, call.operand(1));
+        assert !to.isNullable();
+
+        // Handle NULL.
+
+        if (SqlUtil.isNullLiteral(operand, false)) {
+            setKnownAndValidatedNodeType(operand, HazelcastTypeFactory.INSTANCE.createSqlType(NULL));
+            derived = HazelcastTypeFactory.INSTANCE.createTypeWithNullability(to, true);
+            setKnownAndValidatedNodeType(expression, derived);
+            return derived;
+        }
+
+        derived = to;
+
+        // Assign type for parameters.
+
+        if (isParameter(operand)) {
+            from = HazelcastTypeFactory.INSTANCE.createTypeWithNullability(to, true);
+        }
+
+        // Assign type to numeric literals and validate them.
+
+        BigDecimal numeric = isNumeric(from) || isNumeric(to) ? numericValue(operand) : null;
+
+        if (numeric != null) {
+            from = narrowestTypeFor(numeric, typeName(to));
+        }
+
+        if (isLiteral(operand)) {
+            validateLiteral((SqlLiteral) operand, to);
+        }
+
+        // Infer return type.
+
+        if (isInteger(to) && isInteger(from)) {
+            derived = HazelcastIntegerType.deriveCastType(from, to);
+        } else if (isInteger(to) && numeric != null) {
+            long longValue = numeric.longValue();
+            derived = HazelcastIntegerType.deriveCastType(longValue, to);
+        }
+
+        derived = HazelcastTypeFactory.INSTANCE.createTypeWithNullability(derived, from.isNullable());
+
+        setKnownAndValidatedNodeType(operand, from);
+        setKnownAndValidatedNodeType(expression, derived);
+
+        return derived;
+    }
+
+    private void validateLiteral(SqlLiteral literal, RelDataType type) {
+        SqlTypeName literalTypeName = literal.getTypeName();
+
+        if (!canRepresent(literal, type)) {
+            if (NUMERIC_TYPES.contains(literalTypeName) && isNumeric(type)) {
+                throw newValidationError(literal, RESOURCE.numberLiteralOutOfRange(literal.toString()));
+            } else {
+                throw SqlUtil.newContextException(literal.getParserPosition(),
```

Comment:
AFAIU we lose some error context here. It could be observed it `verify` test results when a conversion from a very large numeric literal causes "numeric overflow" exception, while conversion from the same literal in single quotes causes "cannot convert to BIGINT" exception. I checked Postgres and it is consistent - both values yeilds the same overflow exception,
I think it should be easy for us to fix this - instead of doing a single `canRepresent` check, we may first check if types are convertible (if not - throw `cannot convert`), and only after that perform a separate test for real conversion.

Reply:
created an issue https://github.com/hazelcast/hazelcast/issues/17300 for 4.2